### ISSUE TITLE
MultiGPU training

### ISSUE THREAD

**Issue by littleGiant-28** [herobd/handwriting_line_generation#26]
Hi,
Thanks for creating a clean branch of the code, it has been a great help in understanding the paper. I was wondering if it is possible to train it with multiple GPUS in Pytorch. I tried to do it with, DDP and the code is running but I think the gradients in both GPUs are not in synch (GPU-0 is on iteration-3990 while GPU-1 is on iteration-3975). On further exploration, I found that DDP with retain_graph=True doesn't work in Pytorch. I was wondering if we can recreate the curriculum logic suitable with Pytorch DDP.

**Comment by herobd** [comment]
I don't think there's an easy way using the gadient balancing I'm doing. The documentation says DDP averages on the backwards pass, but I do multiple passes, one for each specific loss.
You could disable the balancing (instead set `loss_weights` in the config to good values, whatever that would be0) and remove the `"no-step"`s from the curriculum and it should run with DDP, but I don't know how the results would turn out.

**Comment by littleGiant-28** [comment]
I also could not think of anything else. I will try some experiments with balancing off.
Thanks.
